File: NodeDemo/loadTests/load-test.js
#!/usr/bin/env node

/**
 * Load Testing Script for Node.js Async/Await Demo
 * Demonstrates the catastrophic impact of blocking the event loop
 */

const http = require('http');

const BASE_URL = 'http://localhost:3000';
const CONCURRENT_REQUESTS = 50;
const TOTAL_REQUESTS = 200;

// Color codes for terminal output
const colors = {
    reset: '\x1b[0m',
    bright: '\x1b[1m',
    red: '\x1b[31m',
    green: '\x1b[32m',
    yellow: '\x1b[33m',
    blue: '\x1b[34m',
    cyan: '\x1b[36m'
};

function log(message, color = colors.reset) {
    console.log(`${color}${message}${colors.reset}`);
}

function makeRequest(path) {
    return new Promise((resolve, reject) => {
        const start = Date.now();
        const url = `${BASE_URL}${path}`;

        http.get(url, (res) => {
            let data = '';

            res.on('data', (chunk) => {
                data += chunk;
            });

            res.on('end', () => {
                const elapsed = Date.now() - start;
                resolve({
                    success: res.statusCode === 200,
                    time: elapsed,
                    status: res.statusCode
                });
            });
        }).on('error', (err) => {
            const elapsed = Date.now() - start;
            resolve({
                success: false,
                time: elapsed,
                status: 'error',
                error: err.message
            });
        });
    });
}

async function checkServer() {
    try {
        await makeRequest('/health');
        return true;
    } catch (error) {
        return false;
    }
}

async function resetMetrics() {
    return new Promise((resolve) => {
        const options = {
            hostname: 'localhost',
            port: 3000,
            path: '/metrics/reset',
            method: 'POST'
        };

        const req = http.request(options, (res) => {
            res.on('data', () => {});
            res.on('end', () => resolve());
        });

        req.on('error', () => resolve());
        req.end();
    });
}

async function runLoadTest(endpoint, name, concurrent = CONCURRENT_REQUESTS, total = TOTAL_REQUESTS) {
    log('\n' + '='.repeat(60), colors.bright);
    log(`Testing: ${name}`, colors.cyan);
    log(`Endpoint: ${endpoint}`, colors.blue);
    log(`Concurrent: ${concurrent} | Total Requests: ${total}`, colors.blue);
    log('='.repeat(60), colors.bright);

    const results = [];
    const batchSize = concurrent;
    const numBatches = Math.ceil(total / batchSize);

    const startTime = Date.now();

    for (let batch = 0; batch < numBatches; batch++) {
        const batchRequests = Math.min(batchSize, total - batch * batchSize);
        process.stdout.write(`  Progress: Batch ${batch + 1}/${numBatches} (${batchRequests} requests)... `);

        const promises = [];
        for (let i = 0; i < batchRequests; i++) {
            promises.push(makeRequest(endpoint));
        }

        const batchResults = await Promise.all(promises);
        results.push(...batchResults);

        log('✓', colors.green);
    }

    const totalTime = (Date.now() - startTime) / 1000;
    const times = results.map(r => r.time).sort((a, b) => a - b);
    const successful = results.filter(r => r.success).length;
    const failed = results.length - successful;

    const stats = {
        totalRequests: results.length,
        successful,
        failed,
        totalTimeSec: totalTime.toFixed(2),
        requestsPerSec: (results.length / totalTime).toFixed(2),
        avgResponseMs: (times.reduce((a, b) => a + b, 0) / times.length).toFixed(2),
        minResponseMs: Math.min(...times).toFixed(2),
        maxResponseMs: Math.max(...times).toFixed(2),
        p50Ms: times[Math.floor(times.length * 0.50)].toFixed(2),
        p95Ms: times[Math.floor(times.length * 0.95)].toFixed(2),
        p99Ms: times[Math.floor(times.length * 0.99)].toFixed(2)
    };

    log('\n' + '─'.repeat(60), colors.bright);
    log('RESULTS:', colors.bright);
    log('─'.repeat(60), colors.bright);
    log(`  Total Requests:     ${stats.totalRequests}`);
    log(`  Successful:         ${stats.successful}`, successful === results.length ? colors.green : colors.yellow);
    log(`  Failed:             ${stats.failed}`, failed > 0 ? colors.red : colors.green);
    log(`  Total Time:         ${stats.totalTimeSec}s`);
    log(`  Throughput:         ${stats.requestsPerSec} req/sec`, colors.cyan);
    log('\n  Response Times:');
    log(`    Average:          ${stats.avgResponseMs} ms`, colors.cyan);
    log(`    Min:              ${stats.minResponseMs} ms`);
    log(`    Max:              ${stats.maxResponseMs} ms`, colors.yellow);
    log(`    P50 (median):     ${stats.p50Ms} ms`);
    log(`    P95:              ${stats.p95Ms} ms`, colors.yellow);
    log(`    P99:              ${stats.p99Ms} ms`, colors.red);
    log('─'.repeat(60), colors.bright);

    return stats;
}

async function main() {
    log('='.repeat(60), colors.bright);
    log('Node.js Async/Await Load Testing', colors.bright);
    log('='.repeat(60), colors.bright);

    // Check server
    log('\nChecking server availability...', colors.yellow);
    const serverRunning = await checkServer();

    if (!serverRunning) {
        log('✗ ERROR: Server is not running at ' + BASE_URL, colors.red);
        log('\nPlease start the server first:', colors.yellow);
        log('  cd NodeDemo', colors.cyan);
        log('  npm install', colors.cyan);
        log('  npm run dev', colors.cyan);
        process.exit(1);
    }

    log('✓ Server is running', colors.green);

    // Reset metrics
    log('\nResetting server metrics...', colors.yellow);
    await resetMetrics();
    log('✓ Metrics reset', colors.green);

    const allResults = {};

    // Test ASYNC endpoints
    log('\n' + '='.repeat(60), colors.bright);
    log('PHASE 1: ASYNC ENDPOINTS (Event Loop Free)', colors.green);
    log('='.repeat(60), colors.bright);

    allResults.asyncUsers = await runLoadTest(
        '/api/async/users',
        'ASYNC - Get Users',
        CONCURRENT_REQUESTS,
        TOTAL_REQUESTS
    );

    allResults.asyncOrders = await runLoadTest(
        '/api/async/orders/123',
        'ASYNC - Get Orders',
        CONCURRENT_REQUESTS,
        TOTAL_REQUESTS
    );

    // Test BLOCKING endpoints
    log('\n' + '='.repeat(60), colors.bright);
    log('PHASE 2: BLOCKING ENDPOINTS (Event Loop Blocked)', colors.red);
    log('⚠️  WARNING: This will freeze the server!', colors.yellow);
    log('='.repeat(60), colors.bright);

    allResults.blockingUsers = await runLoadTest(
        '/api/blocking/users',
        'BLOCKING - Get Users',
        CONCURRENT_REQUESTS,
        TOTAL_REQUESTS
    );

    allResults.blockingOrders = await runLoadTest(
        '/api/blocking/orders/123',
        'BLOCKING - Get Orders',
        CONCURRENT_REQUESTS,
        TOTAL_REQUESTS
    );

    // Comparison
    log('\n' + '='.repeat(60), colors.bright);
    log('COMPARISON: ASYNC vs BLOCKING', colors.bright);
    log('='.repeat(60), colors.bright);

    const comparisons = [
        ['Users Endpoint', 'asyncUsers', 'blockingUsers'],
        ['Orders Endpoint', 'asyncOrders', 'blockingOrders']
    ];

    for (const [name, asyncKey, blockingKey] of comparisons) {
        const asyncStats = allResults[asyncKey];
        const blockingStats = allResults[blockingKey];

        log(`\n${name}:`, colors.cyan);
        log(`  Throughput:`);
        log(`    ASYNC:    ${asyncStats.requestsPerSec} req/sec`, colors.green);
        log(`    BLOCKING: ${blockingStats.requestsPerSec} req/sec`, colors.red);
        const throughputImprovement = ((parseFloat(asyncStats.requestsPerSec) - parseFloat(blockingStats.requestsPerSec)) / parseFloat(blockingStats.requestsPerSec) * 100).toFixed(1);
        log(`    ASYNC is ${throughputImprovement}% faster`, throughputImprovement > 0 ? colors.green : colors.red);

        log(`\n  Average Response Time:`);
        log(`    ASYNC:    ${asyncStats.avgResponseMs} ms`, colors.green);
        log(`    BLOCKING: ${blockingStats.avgResponseMs} ms`, colors.red);
        const responseImprovement = ((parseFloat(blockingStats.avgResponseMs) - parseFloat(asyncStats.avgResponseMs)) / parseFloat(blockingStats.avgResponseMs) * 100).toFixed(1);
        log(`    ASYNC is ${responseImprovement}% faster`, responseImprovement > 0 ? colors.green : colors.red);

        log(`\n  P95 Response Time:`);
        log(`    ASYNC:    ${asyncStats.p95Ms} ms`, colors.green);
        log(`    BLOCKING: ${blockingStats.p95Ms} ms`, colors.red);
    }

    log('\n' + '='.repeat(60), colors.bright);
    log('KEY TAKEAWAYS:', colors.bright);
    log('='.repeat(60), colors.bright);
    log(`
1. ${colors.green}ASYNC${colors.reset} keeps the event loop FREE - handles multiple requests concurrently
2. ${colors.red}BLOCKING${colors.reset} freezes the event loop - processes ONE request at a time
3. In Node.js, blocking = ${colors.red}DISASTER${colors.reset} (single-threaded event loop)
4. Always use async/await for I/O operations (DB, APIs, files)

${colors.yellow}Why is blocking SO BAD in Node.js?${colors.reset}
- Node.js has a SINGLE event loop thread
- Blocking that thread = entire server is frozen
- No other requests can be processed AT ALL
- This is why async/await is CRITICAL in Node.js

${colors.cyan}View detailed metrics at: http://localhost:3000/metrics${colors.reset}
`, colors.reset);
    log('='.repeat(60), colors.bright);
}

main().catch(console.error);

File: CSharpDemo/LoadTests/load-test.py
#!/usr/bin/env python3
"""
Advanced Load Testing Script for Async/Await Demo
Uses concurrent requests to demonstrate async vs sync performance differences
"""

import requests
import time
import statistics
from concurrent.futures import ThreadPoolExecutor, as_completed
import sys

BASE_URL = "http://localhost:5000"
CONCURRENT_REQUESTS = 100
TOTAL_REQUESTS = 500

def check_server():
    """Check if the server is running"""
    try:
        response = requests.get(f"{BASE_URL}/metrics", timeout=5)
        return response.status_code == 200
    except:
        return False

def reset_metrics():
    """Reset server metrics"""
    try:
        requests.post(f"{BASE_URL}/metrics/reset", timeout=5)
        print("✓ Metrics reset")
    except:
        print("✗ Failed to reset metrics")

def make_request(url):
    """Make a single request and measure response time"""
    start = time.time()
    try:
        response = requests.get(url, timeout=30)
        elapsed = (time.time() - start) * 1000  # Convert to milliseconds
        return {
            'success': response.status_code == 200,
            'time': elapsed,
            'status': response.status_code
        }
    except Exception as e:
        elapsed = (time.time() - start) * 1000
        return {
            'success': False,
            'time': elapsed,
            'status': 'error',
            'error': str(e)
        }

def run_load_test(endpoint, name, concurrent=CONCURRENT_REQUESTS, total=TOTAL_REQUESTS):
    """Run load test on a specific endpoint"""
    print(f"\n{'='*60}")
    print(f"Testing: {name}")
    print(f"Endpoint: {endpoint}")
    print(f"Concurrent: {concurrent} | Total Requests: {total}")
    print(f"{'='*60}")

    url = f"{BASE_URL}{endpoint}"
    results = []
    errors = 0

    start_time = time.time()

    # Use ThreadPoolExecutor for concurrent requests
    with ThreadPoolExecutor(max_workers=concurrent) as executor:
        # Submit all requests
        futures = [executor.submit(make_request, url) for _ in range(total)]

        # Collect results as they complete
        completed = 0
        for future in as_completed(futures):
            result = future.result()
            results.append(result['time'])
            if not result['success']:
                errors += 1

            completed += 1
            if completed % 50 == 0:
                print(f"  Progress: {completed}/{total} requests completed")

    total_time = time.time() - start_time

    # Calculate statistics
    if results:
        results.sort()
        stats = {
            'total_requests': total,
            'successful': total - errors,
            'failed': errors,
            'total_time_sec': round(total_time, 2),
            'requests_per_sec': round(total / total_time, 2),
            'avg_response_ms': round(statistics.mean(results), 2),
            'min_response_ms': round(min(results), 2),
            'max_response_ms': round(max(results), 2),
            'p50_ms': round(results[int(len(results) * 0.50)], 2),
            'p95_ms': round(results[int(len(results) * 0.95)], 2),
            'p99_ms': round(results[int(len(results) * 0.99)], 2),
        }

        print(f"\n{'─'*60}")
        print("RESULTS:")
        print(f"{'─'*60}")
        print(f"  Total Requests:     {stats['total_requests']}")
        print(f"  Successful:         {stats['successful']}")
        print(f"  Failed:             {stats['failed']}")
        print(f"  Total Time:         {stats['total_time_sec']}s")
        print(f"  Throughput:         {stats['requests_per_sec']} req/sec")
        print(f"\n  Response Times:")
        print(f"    Average:          {stats['avg_response_ms']} ms")
        print(f"    Min:              {stats['min_response_ms']} ms")
        print(f"    Max:              {stats['max_response_ms']} ms")
        print(f"    P50 (median):     {stats['p50_ms']} ms")
        print(f"    P95:              {stats['p95_ms']} ms")
        print(f"    P99:              {stats['p99_ms']} ms")
        print(f"{'─'*60}")

        return stats
    else:
        print("  ✗ No successful requests")
        return None

def main():
    """Main execution"""
    print("="*60)
    print("Async/Await Load Testing - Advanced")
    print("="*60)

    # Check if server is running
    print("\nChecking server availability...")
    if not check_server():
        print("✗ ERROR: Server is not running at", BASE_URL)
        print("\nPlease start the server first:")
        print("  cd CSharpDemo/AsyncAwaitWebApi")
        print("  dotnet run")
        sys.exit(1)

    print("✓ Server is running")

    # Reset metrics
    print("\nResetting server metrics...")
    reset_metrics()

    # Store results for comparison
    all_results = {}

    # Test ASYNC endpoints
    print("\n" + "="*60)
    print("PHASE 1: ASYNC ENDPOINTS")
    print("="*60)

    all_results['async_users'] = run_load_test(
        "/api/async/users",
        "ASYNC - Get Users",
        concurrent=CONCURRENT_REQUESTS,
        total=TOTAL_REQUESTS
    )

    all_results['async_orders'] = run_load_test(
        "/api/async/orders/123",
        "ASYNC - Get Orders",
        concurrent=CONCURRENT_REQUESTS,
        total=TOTAL_REQUESTS
    )

    # Test SYNC endpoints
    print("\n" + "="*60)
    print("PHASE 2: SYNC ENDPOINTS")
    print("="*60)

    all_results['sync_users'] = run_load_test(
        "/api/sync/users",
        "SYNC - Get Users",
        concurrent=CONCURRENT_REQUESTS,
        total=TOTAL_REQUESTS
    )

    all_results['sync_orders'] = run_load_test(
        "/api/sync/orders/123",
        "SYNC - Get Orders",
        concurrent=CONCURRENT_REQUESTS,
        total=TOTAL_REQUESTS
    )

    # Comparison
    print("\n" + "="*60)
    print("COMPARISON: ASYNC vs SYNC")
    print("="*60)

    comparisons = [
        ('Users Endpoint', 'async_users', 'sync_users'),
        ('Orders Endpoint', 'async_orders', 'sync_orders')
    ]

    for name, async_key, sync_key in comparisons:
        if all_results.get(async_key) and all_results.get(sync_key):
            async_stats = all_results[async_key]
            sync_stats = all_results[sync_key]

            print(f"\n{name}:")
            print(f"  Throughput:")
            print(f"    ASYNC: {async_stats['requests_per_sec']} req/sec")
            print(f"    SYNC:  {sync_stats['requests_per_sec']} req/sec")
            improvement = ((async_stats['requests_per_sec'] - sync_stats['requests_per_sec'])
                          / sync_stats['requests_per_sec'] * 100)
            print(f"    Improvement: {improvement:+.1f}%")

            print(f"\n  Average Response Time:")
            print(f"    ASYNC: {async_stats['avg_response_ms']} ms")
            print(f"    SYNC:  {sync_stats['avg_response_ms']} ms")
            improvement = ((sync_stats['avg_response_ms'] - async_stats['avg_response_ms'])
                          / sync_stats['avg_response_ms'] * 100)
            print(f"    Faster by: {improvement:+.1f}%")

            print(f"\n  P95 Response Time:")
            print(f"    ASYNC: {async_stats['p95_ms']} ms")
            print(f"    SYNC:  {sync_stats['p95_ms']} ms")

    print("\n" + "="*60)
    print("KEY TAKEAWAYS:")
    print("="*60)
    print("""
1. ASYNC endpoints handle MORE concurrent requests with BETTER response times
2. ASYNC releases threads during I/O, allowing better resource utilization
3. SYNC blocks threads, leading to thread pool starvation under load
4. For I/O-bound operations (DB, APIs), ASYNC is significantly better

See the server metrics at: http://localhost:5000/metrics
    """)
    print("="*60)

if __name__ == "__main__":
    main()
